magpie-aws/src/main/java/io/openraven/magpie/plugins/aws/discovery/services/KMSDiscovery.java

```diff
@@ -0,0 +1,139 @@
+import software.amazon.awssdk.core.exception.SdkClientException;
+import software.amazon.awssdk.core.exception.SdkServiceException;
+import software.amazon.awssdk.regions.Region;
+import software.amazon.awssdk.services.kms.KmsClient;
+import software.amazon.awssdk.services.kms.model.*;
+import software.amazon.awssdk.services.kms.model.Tag;
+
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collector;
+import java.util.stream.Collectors;
+
+import static io.openraven.magpie.plugins.aws.discovery.AWSUtils.getAwsResponse;
+import static java.util.Arrays.asList;
+
+public class KMSDiscovery implements AWSDiscovery {
+  private final List<LocalDiscovery> discoveryMethods = asList(
+    this::discoverKeyRotation,
+    this::discoverAliases,
+    this::discoverKeyPolicies,
+    this::discoverGrants,
+    this::discoverTags
+  );
+
+  @FunctionalInterface
+  interface LocalDiscovery {
+    void discover(KmsClient client, KeyListEntry resource, ObjectNode data, Logger logger, ObjectMapper mapper);
+  }
+
+  @Override
+  public void discover(ObjectMapper mapper, Session session, Region region, Emitter emitter, Logger logger) {
+    logger.info("Discovering KMS instances in {}", region);
+    final var client = KmsClient.builder().region(region).build();
+
+    try {
+      client.listKeysPaginator().keys().forEach(key -> {
+        var data = mapper.createObjectNode();
+        data.putPOJO("configuration", key.toBuilder());
+        data.put("region", region.toString());
+
+        for (var dm : discoveryMethods) {
+          try {
+            dm.discover(client, key, data, logger, mapper);
+          } catch (SdkServiceException | SdkClientException ex) {
+            logger.error("Failed to discover data for {}", key.keyArn(), ex);
+          }
+        }
+
+        emitter.emit(new MagpieEnvelope(session, List.of(AWSDiscoveryPlugin.ID + ":kms"), data));
+      });
+    } catch (SdkServiceException | SdkClientException ex) {
+      logger.error("Failed to discover data in {}", region, ex);
+    }
+
+    logger.info("Finished KMS instances discovery in {}", region);
+  }
+
+  private void discoverKeyRotation(KmsClient client, KeyListEntry resource, ObjectNode data, Logger logger, ObjectMapper mapper) {
+    logger.info("Getting Rotation for {}", resource.keyArn());
+    final String keyname = "Rotation";
+    getAwsResponse(
+      () -> client.getKeyRotationStatus(GetKeyRotationStatusRequest.builder().keyId(resource.keyId()).build()),
+      (resp) -> AWSUtils.update(data, Map.of(keyname, resp)),
+      (noresp) -> AWSUtils.update(data, Map.of(keyname, noresp))
+    );
+  }
+
+  private void discoverAliases(KmsClient client, KeyListEntry resource, ObjectNode data, Logger logger, ObjectMapper mapper) {
+    logger.info("Getting Aliases for {}", resource.keyArn());
+    final String keyname = "Aliases";
+    getAwsResponse(
+      () -> client.listAliases(ListAliasesRequest.builder().keyId(resource.keyId()).build()),
```

Comment:
Here it's possible to use listAliasesPaginator. 
Is it needed?
If so how can I use it with AWSUtils.update()?

Reply:
Good question. See my commit.  You've got to stream the paginator results and map them with `.toBuilder()`.